File: Pinball App 2/Pinball App 2/library/LibrarySearchSupport.swift
import Foundation

nonisolated func normalizedSearchTokens(_ value: String) -> [String] {
    value
        .folding(options: [.diacriticInsensitive], locale: .current)
        .lowercased()
        .components(separatedBy: CharacterSet.alphanumerics.inverted)
        .filter { !$0.isEmpty }
}

nonisolated func matchesSearchQuery(_ query: String, fields: [String?]) -> Bool {
    let queryTokens = normalizedSearchTokens(query)
    guard !queryTokens.isEmpty else { return true }

    let haystackTokens = fields.flatMap { normalizedSearchTokens($0 ?? "") }
    guard !haystackTokens.isEmpty else { return false }

    return matchesSearchTokens(queryTokens, haystackTokens: haystackTokens)
}

nonisolated func matchesSearchTokens(_ queryTokens: [String], haystackTokens: [String]) -> Bool {
    guard !queryTokens.isEmpty else { return true }
    guard !haystackTokens.isEmpty else { return false }
    return queryTokens.allSatisfy { queryToken in
        haystackTokens.contains { haystackToken in
            haystackToken.contains(queryToken)
        }
    }
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/library/LibrarySearchSupport.kt
package com.pillyliu.pinprofandroid.library

import java.text.Normalizer

internal fun normalizedSearchTokens(value: String): List<String> {
    val folded = Normalizer.normalize(value, Normalizer.Form.NFD)
        .replace(Regex("\\p{M}+"), "")
    return folded
        .lowercase()
        .split(Regex("[^a-z0-9]+"))
        .filter { it.isNotBlank() }
}

internal fun matchesSearchQuery(query: String, fields: Iterable<String?>): Boolean {
    val queryTokens = normalizedSearchTokens(query)
    if (queryTokens.isEmpty()) return true

    val haystackTokens = fields.flatMap { normalizedSearchTokens(it.orEmpty()) }
    if (haystackTokens.isEmpty()) return false

    return matchesSearchTokens(queryTokens, haystackTokens)
}

internal fun matchesSearchTokens(queryTokens: List<String>, haystackTokens: List<String>): Boolean {
    if (queryTokens.isEmpty()) return true
    if (haystackTokens.isEmpty()) return false
    return queryTokens.all { queryToken ->
        haystackTokens.any { haystackToken -> haystackToken.contains(queryToken) }
    }
}
